Wrap navigation editing features with filters

draganescu opened the pull request:
Advances #25233 
Fixes #25508

## Description
This PR makes the Navigation block's edit component customizable for the Navigation editor. It adds all the formatting controls, the modal list view control, the alignment control in filters and removes them in the Navigation editor.

## How has this been tested?
Tested locally.

## Types of changes
Non breaking changes, but new pattern that may have some effects I am not aware of.


## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/native-mobile.md -->

draganescu committed:
wrap navigation editing features with filters

talldan commented on packages/block-library/src/navigation/edit.js:
```diff
@@ -119,132 +84,247 @@ function Navigation( {
+			<>
+				<BlockEdit { ...props } />
+				<BlockControls>
+					<ToolbarGroup>{ navigatorToolbarButton }</ToolbarGroup>
```
Hopefully it'll be possible to replace this with https://github.com/WordPress/gutenberg/pull/25034

talldan commented on packages/block-library/src/navigation/edit.js:
```diff
@@ -33,64 +34,28 @@ import BlockColorsStyleSelector from './block-colors-selector';
 import * as navIcons from './icons';
 import NavigationPlaceholder from './placeholder';
 
+const ref = createRef();
```
The creates a single `ref` variable, but there can be multiple instances of a block, so not sure how this could work reliably.

kevin940726 reviewed (commented):
I would probably split them into different files `src/navigation/with-inspector-controls.js`, `src/navigation/with-block-controls.js`, etc. But it's fine either way 😆.

talldan commented on packages/block-library/src/navigation/edit.js:
```diff
@@ -119,132 +84,247 @@ function Navigation( {
+);
+
+addFilter(
+	'navigation.BlockEdit',
```
Could these features being implemented using the existing `editor.BlockEdit` filter? It looks like all that's needed to achieve that is to add the following early return to the start of the filter function and it would work:
```
		if ( props.name !== 'core/navigation' ) {
			return <BlockEdit { ...props } />;
		}
```

talldan reviewed (commented):
This is pretty clever. I would personally be happy with this approach if it could be implemented in such a way that it didn't need a new `navigation.BlockEdit` filter, as that constitutes a very specific public API for a single block. In my quick testing it should be possible to use the existing `editor.BlockEdit` filter. I added a comment below on that.

Also noting the item justification isn't working for me in master, but in this PR it's fixed 😄

talldan committed:
Switch to using BlockEdit filter

Change navigation block to use __experimentalColor support option, and disable feature on navigation screen

Remove unused ref

Refactor to use feature flags injected as props to BlockEdit

Also remove font size support

Remove CSS Class and anchor support on navigation screen

talldan commented:
@draganescu let me take over this PR, so I've made some changes:

- Changed the nav block to use the `__experimentalColors` supports property so it's consistent with other blocks. This moves the color settings from the toolbar to the inspector and results in quite a bit of code deletion 🎉 
- Instead of adding features using filters and removing those filters I've done the inverse—added flag props to the navigation block edit that default to true, and then set them to false on the navigation screen using the `BlockEdit` filter. This solves an issue with the previous approach where the features were displayed while the block was still in its initial placeholder state.
- I've used the `registerBlockType` filter to remove `__experimentalColors` and `__experimentalFontSize` support on the navigation screen.

Not sure if there are other options to remove. I wasn't sure about 'Anchor', is that in use?

draganescu commented:
Anchor for the Navigation block doesn't make much sense so we can remove that in all the places.

talldan commented:
I've removed support for both anchor and additional class names on the nav screen for the nav block after a brief consult with @draganescu.

draganescu reviewed (commented):
This works great @talldan

adamziel commented on packages/edit-navigation/src/index.js:
```diff
@@ -31,6 +32,43 @@ function disableInsertingNonNavigationBlocks( settings, name ) {
 	return settings;
 }
 
+function removeNavigationBlockSettingsUnsupportedFeatures( settings, name ) {
+	if ( name !== 'core/navigation' ) {
+		return settings;
+	}
+
+	return {
+		...settings,
+		supports: {
+			...omit( settings.supports, [
+				'anchor',
+				'customClassName',
+				'__experimentalColor',
+				'__experimentalFontSize',
+			] ),
+			customClassName: false,
+		},
+	};
+}
```
I would use an allowlist, not denylist here. With an allowlist, worst case scenario is not offering a new feature on this screen. With a denylist, worst case scenario is breaking the screen with unsupported feature that was added without considering this editor.

talldan replied:
I'm not sure that would be possible as block `supports` settings are an already defined public API.

> With a denylist, worst case scenario is breaking the screen with unsupported feature that was added without considering this editor.

I'm not sure it'd break anything, it'd be a setting that would never have its value saved.

talldan replied:
I suppose the option would be to use a filter to add them to the post/site editor at registration, though I personally feel that blocks in the block library shouldn't require any extra configuration, they should work as intended out of the box.

adamziel commented on packages/edit-navigation/src/index.js:
```diff
@@ -31,6 +32,43 @@ function disableInsertingNonNavigationBlocks( settings, name ) {
+	};
+}
+
+const removeNavigationBlockEditUnsupportedFeatures = createHigherOrderComponent(
+	( BlockEdit ) => ( props ) => {
+		if ( props.name !== 'core/navigation' ) {
+			return <BlockEdit { ...props } />;
+		}
+
+		return (
+			<BlockEdit
+				{ ...props }
+				hasSubmenuIndicatorSetting={ false }
+				hasItemJustificationControls={ false }
+				hasListViewModal={ false }
+			/>
+		);
+	},
+	'removeNavigationBlockEditUnsupportedFeatures'
+);
```
Ditto https://github.com/WordPress/gutenberg/pull/25329/files#r490875716

talldan replied:
I don't agree as there are more cases where the setting needs to be `true` (post editor, site editor) than there are where it needs to be `false` (navigation editor). There are possibilities for mistakes either way.

noisysocks reviewed (approved):
I like the approach! Works well in testing too.
